src/fmnoise/coldbrew.clj

```diff
@@ -1,38 +1,57 @@
 (ns fmnoise.coldbrew
-  (:import (com.github.benmanes.caffeine.cache CacheLoader LoadingCache Caffeine)
+  (:import (com.github.benmanes.caffeine.cache Cache CacheLoader LoadingCache Caffeine)
            (java.time Duration)))
 
 (defn cached
   "Accepts a function and creates cached version of it which uses Caffeine Loading Cache.
-  Cache expiration and refresh time (in seconds) can be provided as `:expire` and `:refresh` meta to function.
-  Example:
-  ```
+  Cache options should can be provided as meta to function. Supported options are:
+  `:expire` - expiration time (in seconds)
+  `:refresh` - refresh time (in seconds)
+  `:when` - function for checking if value should be cached
+
+  Examples:
   (cached ^{:expire 86400 :refresh 36000}
    (fn [db id] (query db id)))
-  ```"
+
+  ;; calculate todays order capacity until it reaches zero
+  (cached ^{:when zero?}
+   (fn [db worker-id] (order-capacity db worker-id)))
+  "
   [f]
-  (let [cache-builder (Caffeine/newBuilder)
+  (let [^Caffeine cache-builder (Caffeine/newBuilder)
+        condition-fn (-> f meta :when)
         _ (when-let [expire (-> f meta :expire)]
             (.expireAfterWrite cache-builder (Duration/ofSeconds expire)))
         _ (when-let [refresh (-> f meta :refresh)]
             (.refreshAfterWrite cache-builder (Duration/ofSeconds refresh)))
-        cache (.build cache-builder (reify CacheLoader (load [_ key] (apply f key))))]
-    (fn [& args] (.get ^LoadingCache cache args))))
+        cache (if condition-fn
+                (.build cache-builder)
+                (.build cache-builder (reify CacheLoader (load [_ key] (apply f key)))))]
+    (if condition-fn
+      (fn [& args]
+        (let [key (or args ::nil)]
+          (or (.getIfPresent ^Cache cache key)
+              (let [computed (apply f args)]
+                (when (condition-fn computed)
+                  (.put ^Cache cache key computed))
+                computed))))
```

Comment:
Or as in [manual population doc](https://github.com/ben-manes/caffeine/wiki/Population#manual) perform something like:
```clj
(let [cache-loader (reify CacheLoader (load [_ key] (apply f key))))]
(or (.getIfPresent ^Cache cache key)
      (let [value (.get ^Cache cache cache-loader)]
        (when (condition-fn value)
          (.put ^Cache cache key value))
```
To perform on the caching library side as much as possible - in case if they handle async `.get` as single computation(as an old caching library they should?)